SDK_versioned_docs/version-3.0.0/guides/auto-router/swap-and-add.md

```diff
@@ -0,0 +1,155 @@
+---
+id: swap-and-add
+title: Swap and Add Liquidity Atomically
+sidebar_position: 2
+---
+
+# Swap and Add Liquidity Atomically
+
+When adding liquidity to a Uniswap v3 pool, you must provide two assets in a particular ratio. In many cases, your contract or the user's wallet hold a different ratio of those two assets. In order to deposit 100% of your assets, you must first **swap** your assets to the optimal ratio and then **add liquidity**. However, the swap may shift the balance of the pool and thus change the optimal ratio!
+
+This guide will teach you how to execute this swap-and-add operation in a single atomic transaction. First, you will use the Auto Router to fetch calldata to swap to the optimal ratio and add liquidity. Then you will submit the transaction to the on-chain router contract `SwapRouter02.sol`. 
+
+## Initializing the Alpha Router
+
+First, [import](https://docs.uniswap.org/sdk/guides/auto-router/quick-start#importing-the-package) and [initialize](https://docs.uniswap.org/sdk/guides/auto-router/quick-start#initializing-the-alpharouter) an instance of the Alpha Router. If you are using a different network, be sure to specify the right `chainId` parameter.
+
+```
+import { AlphaRouter } from "@uniswap/smart-order-router";
+const router = new AlphaRouter({ chainId: 1, provider: web3Provider });
+```
+
+
+## Fetching calldata from `routeToRatio`
+
+Now call the `routeToRatio` method on the Auto Router. This function will return all the calldata you need to submit an atomic swap-and-add transaction.
+
+#### Parameters
+
```

Comment:
I whipped up a markdown table for the parameters! will it work?

```suggestion
| Name                | Requirement | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|---------------------|-------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `token0Balance`     | required    | The initial balance of `token0` that you wish to swap-and-add, where `token0` is the `token0` in your target liquidity pool.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `token1Balance`     | required    | The initial balance of `token1` that you wish to swap-and-add, where `token1` is the `token1` in your target liquidity pool.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `position`          | required    | A [Position](https://docs.uniswap.org/sdk/guides/liquidity/minting#creating-a-position-instance) object that contains the details of the position for which to add liquidity. The position liquidity can be set to `1`, since liquidity is still unknown and will be set inside the call to `routeToRatio`.                                                                                                                                                                                                                                                                                                                                    |
| `swapAndAddConfig`  | required    | A [swapAndAddConfig](https://github.com/Uniswap/smart-order-router/blob/b26ffdc978ab1076c817392ab20ed2df325daf7a/src/routers/router.ts#L123) sets configurations for the `routeToRatio` algorithm. `ratioErrorTolerance` determines the margin of error the resulting ratio can have from the optimal ratio. `maxIterations` determines the maximum times the algorithm will iterate to find a ratio within error tolerance. If `maxIterations` is exceeded, an error is returned.                                                                                                                                                             |
| `swapAndAddOptions` | optional    | If [swapAndAddOptions](https://github.com/Uniswap/smart-order-router/blob/b26ffdc978ab1076c817392ab20ed2df325daf7a/src/routers/router.ts#L130) is included, `routeToRatio` will return the calldata for executing the atomic swap-and-add. These options contain `swapConfig` and `addLiquidityOptions`. `swapConfig` configures to set a recipient of leftover dust from swap, slippageTolerance, deadline, inputTokenPermit and outputTokenPermit. `addLiquidityOptions` must contain a `tokenId` to add to an existing position, or `recipient` to mint a new one. It also includes a slippage tolerance and deadline for adding liquidity. |
| `routingConfig`     | optional    | Optional advanced config for tuning the performance of the routing algorithm. View the [AlphaRouterConfig](https://github.com/Uniswap/smart-order-router/blob/b26ffdc978ab1076c817392ab20ed2df325daf7a/src/routers/alpha-router/alpha-router.ts#L222) object for these optional configuration parameters.                                                                                                                                                                                                                                                                                                                                      |
```